include/seqan3/io/sequence/sequence_file_in.hpp

```diff
@@ -14,155 +58,123 @@ namespace seqan3
-        std::function<void(std::string &)>  sequence_filter = [] (std::string & seq)  {};
-        std::function<void(std::string &)>   id_filter = [] (std::string & id)   {};
+        std::function<void(std::string &)> sequence_filter = [] (std::string & seq) {};
+        std::function<void(std::string &)> meta_filter = [] (std::string & meta) {};
         std::function<void(std::string &)> qual_filter = [] (std::string & qual) {};
     };
     options_type options;
 
-protected:
-    ~sequence_file_in() = default;
-
-private:
-    /* file format */
-    std::string file_name;
-    stream_type stream;
-    valid_formats format;
+    /* member functions */
+    // TODO make the requirements stricter
+    template <typename sequence_type, typename meta_type, typename qual_type>
+        requires sequence_concept<std::decay_t<sequence_type>> &&
+                 sequence_concept<std::decay_t<meta_type>> &&
+                 sequence_concept<std::decay_t<qual_type>>
+    inline void read(sequence_type && seq,
+                     meta_type && meta = std::string{},
+                     qual_type && qual = std::string{});
+
+    template <typename seqs_type, typename metas_type, typename quals_type>
+        requires sequence_of_sequence_concept<std::decay_t<seqs_type>> &&
+                 sequence_of_sequence_concept<std::decay_t<metas_type>> &&
+                 sequence_of_sequence_concept<std::decay_t<quals_type>>
+    inline void read(seqs_type && seqs,
+                     metas_type && metas = std::vector<std::string>{},
+                     quals_type && quals = std::vector<std::string>{},
+                     size_t max_records = 0);
 
-    /* private functions */
-    void select_decompression(std::string const & compress_ext);
-    template <size_t index>
-    void assign_format(std::string const & ext);
-
-    /* buffers */
+protected:
+    /* member variables */
+    using detail::file_base<sequence_file_in_traits>::format;
+    using detail::file_base<sequence_file_in_traits>::stream;
 };
 
 // ------------------------------------------------------------------
 // public API
 // ------------------------------------------------------------------
 
-sequence_file_in::sequence_file_in(std::string const & _file_name) :
-        file_name(_file_name)
+template <typename sequence_file_in_traits>
+template <typename sequence_type, typename meta_type, typename qual_type>
+    requires sequence_concept<std::decay_t<sequence_type>> &&
+             sequence_concept<std::decay_t<meta_type>> &&
+             sequence_concept<std::decay_t<qual_type>>
+inline void sequence_file_in<sequence_file_in_traits>::read(sequence_type && seq,
+                                                            meta_type && meta,
+                                                            qual_type && qual)
 {
-    // open stream
-    stream.open(_file_name, std::ios::binary);
+    if (format.valueless_by_exception())
+        throw "format not set!";
```

Comment:
Can you add a todo, that we have to change the exception. 
actually we do not want to throw arbitrary types but real exception types.